### PR TITLE
NullPointerException in JobMapper when referencing image on another disk drive.

### PR DESCRIPTION
**Pull request by jlacour31**
## Description
A NPE appears in a topic refers an image from another drive.

## Motivation and Context
Fix for https://github.com/dita-ot/dita-ot/issues/3635

## How Has This Been Tested?
I added a new assert in CleanPreprocessModuleTest.getCommonBase()

## Type of Changes
- Bug fix _(non-breaking change which fixes an issue)_

## Checklist
- My code follows the code style of this project.
    -  <https://github.com/dita-ot/dita-ot/wiki/Java-Coding-Conventions>
    -  <https://github.com/dita-ot/docs/wiki/coding-guidelines>
- I have updated the unit tests to reflect the changes in my code.

### CODE REVIEW

File: src/test/java/org/dita/dost/module/CleanPreprocessModuleTest.java [hunk 1]
```diff
@@ -52,6 +52,7 @@ public void getCommonBase() {
         assertEquals(create("file:/foo/"), module.getCommonBase(create("file:/foo/a"), create("file:/foo/bar/b")));
         assertEquals(create("file:/foo/"), module.getCommonBase(create("file:/foo/bar/a"), create("file:/foo/b")));
         assertEquals(create("file:/foo/"), module.getCommonBase(create("file:/foo/bar/a"), create("file:/foo/baz/b")));
+        assertEquals(create("file:/C:/"), module.getCommonBase(create("file:/C:/a"), create("file:/D:/b")));
```
Comment (jelovirt) [inline src/test/java/org/dita/dost/module/CleanPreprocessModuleTest.java:55]: If you revert the code change, this test doesn't test for NPE.

**Review by jelovirt — CHANGES_REQUESTED** [review]

File: src/main/java/org/dita/dost/module/CleanPreprocessModule.java [hunk 2]
```diff
@@ -263,11 +263,12 @@ URI getCommonBase(final URI left, final URI right) {
                 //
             } else {
                 final int common = Math.max(0, i);
-                final String path = Arrays.asList(la)
-                        .subList(0, common)
-                        .stream()
-                        .collect(Collectors.joining("/")) + "/";
-                return URLUtils.setPath(left, path);
+                final List<String> commons = Arrays.asList(la).subList(0, common);
+                if (commons != null && commons.size() > 1) {
```
Comment (jelovirt) [inline src/main/java/org/dita/dost/module/CleanPreprocessModule.java:267]: `common` can never be null.

**Comment by jelovirt** [comment]
@jlacour31 could you attach a stacktrace of the the NPE that this fixes and a test case, because I can't reproduce a NPE with the tests that were added.

File: src/main/java/org/dita/dost/module/CleanPreprocessModule.java (hunk 2 above)
Comment (jelovirt) [inline src/main/java/org/dita/dost/module/CleanPreprocessModule.java:267]: With a test

```java
assertEquals(create("file:/foo/"), module.getCommonBase(create("file:/foo/a/b/c"), create("file:/bar/b/c/d")));
```

this code passes, but the result is incorrect. The correct results should be `file:/`. The code makes assumption that the first directory in the path token should be retained and the resulting path can never start from root. This only applies to Windows, not *NIX.

**Review by jelovirt — CHANGES_REQUESTED** [review]

File: src/main/java/org/dita/dost/module/CleanPreprocessModule.java (hunk 2 above)
Comment (jlacour31) [reply src/main/java/org/dita/dost/module/CleanPreprocessModule.java]: I removed the null check.

File: src/test/java/org/dita/dost/module/CleanPreprocessModuleTest.java (hunk 1 above)
Comment (jlacour31) [reply src/test/java/org/dita/dost/module/CleanPreprocessModuleTest.java:55]: You're right, this doesn't cover directly the NPE, I don't know where I can test it.

File: src/main/java/org/dita/dost/module/CleanPreprocessModule.java (hunk 2 above)
Comment (jlacour31) [reply src/main/java/org/dita/dost/module/CleanPreprocessModule.java]: The problem is in fact that if the common path is only `file:/` a NPE appears in the HTML5 transformation.
And you're right this problem is only Windows related as Linux doesn't have drives.

**Comment by jlacour31** [comment]
@jelovirt I copied the stacktrace from the original issue posted by @raducoravu:
```
DITA-OT3.x\plugins\org.dita.base\build_preprocess.xml:364: java.lang.NullPointerException
at org.dita.dost.ant.types.JobMapper.mapFileName(JobMapper.java:79)
at org.apache.tools.ant.util.CompositeMapper.lambda$mapFileName$0(CompositeMapper.java:32)
at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)
at java.util.stream.ReferencePipeline$2$1.accept(Unknown Source)
```

**Comment by raducoravu** [comment]
So I'm pretty sure while the fix in this PR will prevent a NPE, it will still result in incorrect output.

@jelovirt well it's not great at all for a process to fail with a NPE, as the end user does not know why it fails, ideally if it fails in such cases it should fail with a precise error message, or show a precise error message and continue to work (probably with a broken link or two somewhere). How about if your validation filter would report such problems as precise error messages? And we also guard the NPE.

**Comment by jelovirt** [comment]
@raducoravu You have a Windows machine. Could you create a small test where you have e.g.

- `c:\foo\root.ditamap`
- `c:\foo\topic.dita`
- `d:\bar\other.dita`

And you run preprocess until cleanup method? So e.g. add `<fail>` as the first step of `clean-preprocess` target. Then, could you paste the `.job.xml` here? I'd like to see what the `FileInfo`s look like in this case.

**Comment by raducoravu** [comment]
@jelovirt should I perform this fix with the patch made by @jlacour31 added or not, just with the default DITA OT 3.7 code?

**Comment by raducoravu** [comment]
@jelovirt I did what you asked, my c:\foo\root.ditamap has the topicrefs like this:

        <map>
            <title></title>
            <topicref href="topic.dita"/>
            <topicref href="file:/D:/bar/other.dita"/>
        </map>

I published to HTML5 using DITA OT 3.6 (which does not have Julien's patch applied). Added a fail before the "clean-preprocess"
Attaching the entire contents of the temp folder:

[html5.zip](https://github.com/dita-ot/dita-ot/files/8213931/html5.zip)

I'm not sure if this test is similar to the current problem, when an image is referenced from another drive though.

**Comment by jelovirt** [comment]
@raducoravu I pushed a fix for Windows to this PR branch. Could you run the unit tests in `CleanPreprocessModuleTest` and check what this fix does with the problematic image on a different drive?

**Comment by jelovirt** [comment]
> following errors in the console at different stages

@raducoravu that's to be expected, and all those preprocess steps are ran _before_ clean-preprocess target. So this fix works as expected because it's a preprocess1 limitation that you cannot mix source from different drives on Windows.

What's the results with the original test set with the image? No NPE from there?

**Comment by raducoravu** [comment]
@jelovirt if I remove the path to the bar/other.dita from the DITA Map and add inside "topic.dita" an absolute image reference like this:

      <image href="file:/D:/projects/eXml/samples/dita/flowers/images/Gardeniaflower.jpg"                 id="image_uv5_zxd_vsb"/>

then publish to HTML5 I get these errors in the console:

     copy-image:     
      [copy] Copying 1 file to C:\foo\out\html5 Attempt to copy C:\foo\D:\projects\eXml\samples\dita\flowers\images\Gardeniaflower.jpg to C:\foo\out\html5\D:\projects\eXml\samples\dita\flowers\images\Gardeniaflower.jpg using NIO Channels failed due to 'The filename, directory name, or volume label syntax is incorrect'.  Falling back to streams.      
       [copy]       Failed to copy C:\foo\D:\projects\eXml\samples\dita\flowers\images\Gardeniaflower.jpg to C:\foo\out\html5\D:\projects\eXml\samples\dita\flowers\images\Gardeniaflower.jpg due to The filename, directory name, or volume label syntax is incorrect

but the publishing continues, no NPE. The result topic.html contains an absolute path to the image file:

**Comment by jelovirt** [comment]
@raducoravu That sounds about right. This fixed the NPE but the code.

**Comment by jelovirt** [comment]
This PR contains a merge commit that should be rebased out, but to save time, I'm merging this as is.

**Pull request merged**